src/transformers/trainer.py

```diff
@@ -1762,6 +1762,11 @@ def _inner_training_loop(
                 else:
                     tr_loss_step = self.training_step(model, inputs)
 
+                # Reduce gradients first for XLA device to be consistent with native pytorch behavior
+                if is_torch_tpu_available():
```

Comment:
This should only be done if gradient scaling is necessary, otherwise it adds computation for nothing.

Reply:
Could you elaborate more?
My understanding is `all_reduce` is necessary when running on multiple devices, regardless of if gradient scaling is used.
This PR essentially modified the sequence of operations from:
```py
loss = loss_fn(...)
loss.backward()
if max_grad_norm > 0:
    if do_grad_scaling > 0:
        xm.all_reduce(...)
        self.scaler.unscale_(self.optimizer)
    nn.utils.clip_grad_norm_(...)
if self.do_grad_scaling:
    self.scaler.step(self.optimizer)
    self.scaler.update()
else:
    xm.optimizer_step(self.optimizer)
```
to
```py
loss = loss_fn(...)
loss.backward()
xm.all_reduce(...)
if max_grad_norm > 0:
    if do_grad_scaling > 0:
        self.scaler.unscale_(self.optimizer)
    nn.utils.clip_grad_norm_(...)
if self.do_grad_scaling:
    self.scaler.step(self.optimizer)
    self.scaler.update()
else:
    self.optimizer.step()
```
Note `xm.optimizer_step() == xm.all_reduce(...) + optimizer.step()` [code](https://github.com/pytorch/xla/blob/28ea3758e9586ef8cc22270a16e1ddb4a21aa6f7/torch_xla/core/xla_model.py#L1021-L1022), here we need to decouple `xm.all_reduce` and `optimizer.step` in favor of the extra gradient clipping step in between.

For single device case, it won't add new computations in the hlo graph due to the flag [here](https://github.com/pytorch/xla/blob/28ea3758e9586ef8cc22270a16e1ddb4a21aa6f7/torch_xla/core/xla_model.py#L584), which evaluates to false for single device case.

Reply:
That's not the exact code that PyTorch XLA executes. I'd rather keep using the code from PyTorch XLA which is maintained than reproduce it here in all cases.